transformers.py

```diff
@@ -0,0 +1,151 @@
+import sonnet as snt
+import tensorflow as tf
+import math
+
+def positional_encoding(seq_len, d_model, batch_size=1):
+    """
+    Returns a tensor following the postional encoding function
+     (sinusodal from Vaswani et. all 2017).
+
+    Return shape: [batch, seq_len, d_model]
+    """
+    def encoding_angle(pos, i):
+        pos = tf.cast(pos, tf.dtypes.float32)
+        i = tf.cast(i, tf.dtypes.float32)
+        denom = tf.math.pow(10000, i/d_model)
+        return pos / denom
+
+    # BUG: Fails to serialize these calls:
+    # "Failed to convert object of type <class 'list'> to Tensor. Contents: [None, 480]. Consider casting elements to a supported type."
+    i_tensor = tf.expand_dims(tf.range(0, d_model//2), 0) # [1, d_model/2]
+    i_tensor = tf.broadcast_to(i_tensor, [seq_len, d_model//2]) # [seq_len, d_model/2]
+    j_tensor = tf.expand_dims(tf.range(0, seq_len), 1) # [seq_len, 1]
+    j_tensor = tf.broadcast_to(j_tensor, [seq_len, d_model//2]) # [seq_len, d_model/2]
+    angles = encoding_angle(j_tensor, i_tensor) # [seq_len, d_model/2]
+
+    # Apply sin to even indices, cos to odd indices
+    sins_angle = tf.math.sin(angles) # [seq_len, d_model/2]
+    coss_angles = tf.math.cos(angles) # [seq_len, d_model/2]
+    joined = tf.concat([sins_angle, coss_angles], -1) # [s, d]
+
+    #Add in batch
+    encoding = tf.expand_dims(joined, 0)
+    encoding = tf.repeat(encoding, [batch_size], axis=0) # [b, s, d]
+    return encoding
+
+def attention(queries, keys, values, masked=True):
+    """
+    Returns the 'attention' between three sequences: keys, queries, and values
+    Specifically this implementation uses 'scaled dot-product' attention.
+
+    This can be seen as a measure of the compatibility or relative importance between the keys and queries.
+    This compatilbility is then applied to the 'input' sequence represented by values.
+
+    Returns a tensor with the same shape as Values where [b, i ,j] represents the relative importance
+    "attention" of element j in the sequence.
+
+    keys: (batch, seq_len, D_k)
+    queries: (batch, seq_len, D_k)
+    values: (batch, seq_len, D_v)
+    returns: (batch, seq_len, D_v)
+    """
+    assert keys.shape == queries.shape, "keys and values must have equivalent shapes"
+    # compat [b, i, j] is the dot product of key i and query j (for batch # b)
+    compat = tf.matmul(queries, tf.transpose(keys, [0, 2, 1])) # [B, S, S]
```

Comment:
Instead of explicitly calling `tf.transpose`, you can set the `transpose_b` argument for `tf.matmul` which should be faster.